sw/otbn/crypto/rsa_keygen.s

```diff
@@ -1316,29 +1286,76 @@ relprime_f4:
 
   /* Isolate the lower 32 bits of the sum.
        w22 <= w23[31:0] */
-  bn.rshi  w22, w23, w31 >> 32
-  bn.rshi  w22, w31, w22 >> 224
+  bn.and   w22, w23, w24 >> 224
 
   /* Add the two halves of the sum (technically 32 and 34 bits). A carry was
      not possible in the previous addition since the value is too small.
        w23 <= (w22 + (w23 >> 32)) */
   bn.add   w23, w22, w23 >> 32
 
-  /* Note: At this point, we're down to the last few terms:
-       x \equiv (w23[15:0] - w23[31:16] + w23[34:32]) mod F4 */
+  ret
+
+/**
+ * Check if a large number is relatively prime to 65537 (aka F4).
+ *
+ * Returns a nonzero value if GCD(x,65537) == 1, and 0 otherwise
+ *
+ * A naive implementation would simply check if GCD(x, F4) == 1, However, we
+ * can simplify the check for relative primality using a few helpful facts
+ * about F4 specifically:
+ *   1. It is prime.
+ *   2. It has the special form (2^16 + 1).
+ *
+ * Because F4 is prime, checking if a number x is relatively prime to F4 means
+ * simply checking if x is a direct multiple of F4; if (x % F4) != 0, then x is
+ * relatively prime to F4. This means that instead of computing GCD, we can use
+ * basic modular arithmetic.
+ *
+ * Here, the special form of F4, fact (2), comes in handy. Since 2^32 mod F4 =
+ * 1, we can use `fold_bignum` to bring the number down to 35 bits cheaply.
+ *
+ * Since 2^16 is equivalent to -1 modulo F4, we can express the resulting
+ * number in base-2^16 and simplify as follows:
+ *   x = x0 + 2^16 * x1 + 2^32 * x2
+ *   x \equiv x0 + (-1) * x1 + (-1)^2 * x2
+ *   x \equiv x0 - x1 + x2 (mod F4)
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  x16: dptr_x, pointer to first limb of x in dmem
+ * @param[in]  x30: n, number of 256-bit limbs for x
+ * @param[in]  w31: all-zero
+ * @param[out] w22: result, 0 only if x is not relatively prime to F4
+ *
+ * clobbered registers: x2, w22, w23
+ * clobbered flag groups: FG0
+ */
+relprime_f4:
+  /* Load F4 into the modulus register for later.
+       MOD <= 2^16 + 1 */
+  bn.addi  w22, w31, 1
+  bn.add   w22, w22, w22 << 16
+  bn.wsrw  MOD, w22
+
+  /* Generate a 256-bit mask.
+       w24 <= 2^256 - 1 */
+  bn.not   w24, w31
+
+  /* Fold the bignum to get a 35-bit number r such that r mod F4 = x mod F4.
+       w23 <= r */
+  jal      x1, fold_bignum
 
   /* Isolate the lower 16 bits of the 35-bit working sum.
        w22 <= w23[15:0] */
-  bn.rshi  w22, w23, w31 >> 16
-  bn.rshi  w22, w31, w22 >> 240
+  bn.and   w22, w23, w24 >> 240
 
   /* Add the lower 16 bits of the sum to the highest 3 bits to get a 17-bit
      result.
        w22 <= w22 + (w23 >> 32) */
   bn.add   w22, w22, w23 >> 32
 
-  /* The sum from the previous addition is < 2 * F4, so a modular addition with
-     zero is sufficient to fully reduce.
+  /* The sum from the previous addition is <= 2^16 - 1 + 2^3 - 1 < 2 * F4, so a
```

Comment:
Maybe "at most" instead of "<=", since those symbols are also being used for assignment?